tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc

```diff
@@ -319,10 +319,10 @@ llvm::Value* EmitFullWarpShuffleDown(llvm::Value* value, llvm::Value* offset,
 
   // Special case for efficiency
   if (value->getType()->isFloatTy() && bit_width == 32) {
-    return EmitCallToTargetIntrinsic(
-        TargetIntrinsicID::kShflDownF32,
-        {all_warps_mask, value, offset, builder->getInt32(kWarpSize - 1)}, {},
-        builder);
+    return EmitCallToTargetFunction(
+        TargetFunctionID::kShflDownF32,
+        {all_warps_mask, value, offset, builder->getInt32(kWarpSize - 1)},
+        {S32, F32, S32, S32}, F32, {}, {}, builder);
```

Comment:
With the addition of these two parameters which are usually the empty list, this API seems difficult to use correctly.

I am wondering whether we should use "pass-by-struct" instead.

```
TargetFnCallInfo info(TargetFunctionId::kShflDownF32, builder);
info.operands = {...};
info.input_types = {...};
info.attributes = {...};
info.overloaded_types = {...};
EmitCallToTargetFunction(info);
```

You could have an overload of EmitCallToTargetFunction which accepts the most commonly-used params so you don't have to use the "call-by-struct" method all the time.

WDYT?